kotlin/workflow-runtime/src/main/java/com/squareup/workflow/internal/SubtreeManager.kt

```diff
@@ -79,15 +82,36 @@ internal class SubtreeManager<StateT, OutputT : Any>(
   /**
    * Uses [selector] to invoke [WorkflowNode.tick] for every running child workflow this instance
    * is managing.
+   *
+   * If one of this workflow's children emits an output, [handler] will be called with the action
+   * assigned to that workflow and a [Source.Subtree] describing the child that triggered the
+   * update.
+   *
+   * Note that if [handler] is called with [Kind.ChildDidUpdate], the workflow action that gets
+   * passed in will always be [noAction]. This is the only logical value because
+   * [Kind.ChildDidUpdate] specifically means that this workflow did not actually get updated
+   * itself, but one of its descendants did and this one is just part of the path down the tree.
    */
   fun <T : Any> tickChildren(
-    selector: SelectBuilder<T?>,
-    handler: (WorkflowAction<StateT, OutputT>) -> T?
+    selector: SelectBuilder<OutputEnvelope<T>>,
+    handler: (WorkflowAction<StateT, OutputT>, Kind) -> OutputEnvelope<T>
   ) {
     for ((case, host) in hostLifetimeTracker.lifetimes) {
       host.tick(selector) { output ->
-        val componentUpdate = case.acceptChildOutput(output)
-        return@tick handler(componentUpdate)
+        if (output.output != null) {
+          val componentUpdate = case.acceptChildOutput(output.output)
+          // This workflow's child emitted a non-null output: the WorkflowNode will execute the
+          // WorkflowAction, and we need to pass DidUpdate to indicate that this workflow actually
+          // got updated.
+          return@tick handler(
```

Comment:
Can you hoist the `return@tick` out to the `if`?

Reply:
Probably